chainer/link_hooks/spectral_normalization.py

```diff
@@ -225,20 +237,34 @@ def _prepare_parameters(self, link, input_variable=None):
         if self.use_gamma:
             # Initialize the scaling parameter with the max singular value.
             weight_matrix = self.reshape_W(initialW.array)
-            _, s, _ = link.xp.linalg.svd(weight_matrix)
+            # TODO(crcrpar): Remove this when chainerx supports SVD.
+            if link.xp is chainerx:
+                xp, device, array = fallback._from_chx(weight_matrix)
+                if xp is numpy:
+                    _, s, _ = numpy.linalg.svd(array)
+                else:
+                    with chainer.using_device(device):
+                        _, s, _ = xp.linalg.svd(array)
+            else:
+                _, s, _ = link.xp.linalg.svd(weight_matrix)
             with link.init_scope():
                 link.gamma = variable.Parameter(s[0], ())
         self._initialized = True
 
-    def normalize_weight(self, link, *args, **kwargs):
+        # N.B.(crcrpar): Tentative adjustment for half precision.
+        if initialW.dtype == numpy.float16:
+            self.eps = 1e-6
```

Comment:
Could you make it documented?  `__init__` can have a default `eps=None` that sets `self.eps = 1e-6` if float16 or `self.eps = 1e-12` otherwise.

Reply:
I discussed it offline, and I found Chainer team reached the consensus (when `eps` of `Adam` is discussed) that hyper-parameters should not depend on precision.

Making `eps=1e-6` default looks good (before v6 release).

Reply:
That makes sense to me.

One thing I'm concerned with is that the original implementation uses 1e-12.
ref: https://github.com/pfnet-research/sngan_projection/blob/master/source/functions/max_sv.py#L5